### PR TITLE
feat: add versionCatalogue method

### LINKED ISSUE #1385: Use version catalogs in generated projects
Currently we use the "good old" dependency coordinates in generated templates. But instead, we could use the version catalog that we publish alongside the BOM.

For example, this is what the dependencies of a project with Micronaut Data and PostgreSQL would look like when using a catalog:

```gradle
dependencies {
    annotationProcessor(mn.micronaut.data.processor)
    annotationProcessor(mn.micronaut.http.validation)
    implementation(mn.micronaut.http.client)
    implementation(mn.micronaut.jackson.databind)
    implementation(mn.micronaut.data.jdbc)
    implementation(mn.micronaut.jdbc.hikari)
    implementation(mn.jakarta.annotation.api)
    runtimeOnly(mn.logback)
    runtimeOnly(mn.postgresql)
    implementation(mn.micronaut.validation)
}
```

Note that this requires updating our generated `settings.gradle` file:

```gradle
dependencyResolutionManagement {
   repositories {
       mavenCentral()
   }
   
   repositoriesMode = RepositoriesMode.PREFER_PROJECT
   
   versionCatalogs {
       create("mn") {
           from("io.micronaut:micronaut-bom:$micronautVersion")
       }
   }
}
```

Also for dependencies which are not in the published catalog, we'd have to add them in the `libs` catalog.

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

File: starter-core/src/main/java/io/micronaut/starter/build/gradle/GradleDependency.java [hunk 1]
```diff
@@ -119,6 +122,14 @@ public String toSnippet() {
         return snippet;
     }
 
+    @NonNull
+    Optional<String> versionCatalogue() {
```
Comment (melix) [inline starter-core/src/main/java/io/micronaut/starter/build/gradle/GradleDependency.java]: I know you like the British version, but I think we should stick with the official terminology which is "version catalog" :)

Comment (sdelamo) [reply starter-core/src/main/java/io/micronaut/starter/build/gradle/GradleDependency.java]: resolved via https://github.com/micronaut-projects/micronaut-starter/pull/1387/commits/0f184501cebbf3992c826e1abf072570ea5e3a6b

**Review by melix — APPROVED** [review]

**Comment by burtbeckwith** [comment]
@sdelamo these changes aren't backwards compatible. I don't think it's a big deal since starter libs aren't likely to be used in user apps, but an app that compiles with 3.6.x will not compile after this change. Consider adding overloads of `GradleBuildCreator.create()` and the `GradleDependency` constructor that pass `false` for `useVersionCatalogue`

**Comment by alvarosanchez** [comment]
It is a big deal. This has broken micronaut-guides with 3.7.

**Comment by sdelamo** [comment]
> @sdelamo these changes aren't backwards compatible. I don't think it's a big deal since starter libs aren't likely to be used in user apps, but an app that compiles with 3.6.x will not compile after this change. Consider adding overloads of `GradleBuildCreator.create()` and the `GradleDependency` constructor that pass `false` for `useVersionCatalogue`

I am not treating starter as other modules. I am introducing binary compatibility changes between minor versions. We can change that decision if we want. But this flexibility allows us to evolve Micronaut Guides faster.

**Pull request merged**